src/btreemap/node/v2.rs

```diff
@@ -0,0 +1,402 @@
+            for _ in 0..num_entries + 1 {
+                let child = address_from_slice(&node_buf, offset);
+                offset += Address::size();
+                children.push(child);
+            }
+        }
+
+        // Load the keys.
+        let mut keys = Vec::with_capacity(num_entries);
+        let mut encoded_values = Vec::with_capacity(num_entries);
+        let mut buf = vec![];
+        for _ in 0..num_entries {
+            // Load the key's size.
+            // TODO: store the size in a more efficient way.
+            // TODO: don't store the size if the key is fixed.
+            let key_size = read_u32_from_slice(&node_buf, offset) as usize;
+            offset += U32_SIZE;
+
+            // Load the key.
+            buf.resize(key_size, 0);
+            let key = K::from_bytes(Cow::Borrowed(
+                &node_buf[offset.get() as usize..offset.get() as usize + key_size],
+            ));
+            offset += Bytes::from(key_size as u64);
+            keys.push(key);
+        }
+
+        // Load the values
+        for _ in 0..num_entries {
+            // Load the value's size.
+            let value_size = read_u32_from_slice(&node_buf, offset) as usize;
+            offset += U32_SIZE;
+
+            // Load the value.
+            // TODO: Read values lazily.
+            buf.resize(value_size, 0);
+            encoded_values.push(Value::ByVal(
+                node_buf[offset.get() as usize..offset.get() as usize + value_size].to_vec(),
+            ));
+
+            offset += Bytes::from(value_size as u64);
+        }
+
+        let original_overflow_address = address_from_slice(&node_buf, OVERFLOW_ADDRESS_OFFSET);
+
+        Self {
+            address,
+            keys,
+            encoded_values: RefCell::new(encoded_values),
+            children,
+            node_type,
+            version: Version::V2(page_size),
+            overflow: if original_overflow_address == crate::types::NULL {
+                None
+            } else {
+                Some(original_overflow_address)
+            },
+        }
+    }
+
+    // Saves the node to memory.
+    pub(super) fn save_v2<M: Memory>(&self, allocator: &mut Allocator<M>) {
+        assert_eq!(self.keys.len(), self.encoded_values.borrow().len());
+
+        let page_size = self.version.page_size();
+
+        // A buffer to serialize the node into first, then write to memory.
+        let mut buf = vec![];
+        buf.extend_from_slice(MAGIC);
+        buf.extend_from_slice(&[LAYOUT_VERSION_2]);
+        buf.push(match self.node_type {
+            NodeType::Leaf => LEAF_NODE_TYPE,
+            NodeType::Internal => INTERNAL_NODE_TYPE,
+        });
+        buf.extend_from_slice(&(self.keys.len() as u16).to_le_bytes());
+
+        // Add a null overflow address. This might get overwritten later in case the node
+        // does overflow.
+        buf.extend_from_slice(&[0; 8]);
```

Comment:
It's probably a low impact change and perhaps you plan to address in follow ups anyway but I'll still ask: can we create a vector with capacity for the first few things we need to add to the buffer? It seems like it should be possible because we should know how much space these take. (The keys and values later need to be done on demand of course).